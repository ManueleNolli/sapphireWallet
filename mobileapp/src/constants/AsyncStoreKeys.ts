const theme = 'theme'
const firstAccess = 'firstAccess'

const EOAAddress = 'EOAAddress'

const walletContractAddress = 'walletContractAddress'

export { theme, firstAccess, EOAAddress, walletContractAddress }
