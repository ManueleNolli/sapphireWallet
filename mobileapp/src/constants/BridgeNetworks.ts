export enum BRIDGE_NETWORKS {
  MUMBAI = 'mumbai',
}
