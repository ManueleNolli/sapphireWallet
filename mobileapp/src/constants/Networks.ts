export enum NETWORKS {
  LOCALHOST = 'Localhost',
  SEPOLIA = 'Sepolia',
}
