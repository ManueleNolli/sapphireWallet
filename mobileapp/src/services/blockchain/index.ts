export * from './utils'
export * from './OwnedNFTs'
