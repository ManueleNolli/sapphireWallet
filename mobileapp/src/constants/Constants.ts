import * as asyncStoreKeys from './AsyncStoreKeys'
import * as secureStoreKeys from './SecureStoreKeys'

export default {
  asyncStoreKeys,
  secureStoreKeys,
}
