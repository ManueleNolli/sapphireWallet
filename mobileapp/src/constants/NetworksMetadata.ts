import { NETWORKS } from './Networks'
import { BRIDGE_NETWORKS } from './BridgeNetworks'

const ALL_NETWORKS = { ...NETWORKS, ...BRIDGE_NETWORKS }

type ChainIdsToNetwork = {
  [key: string]: (typeof ALL_NETWORKS)[keyof typeof ALL_NETWORKS]
}

export const CHAIN_IDS_TO_NETWORK: ChainIdsToNetwork = {
  '1337': ALL_NETWORKS.LOCALHOST,
  '11155111': ALL_NETWORKS.SEPOLIA,
  '80001': ALL_NETWORKS.MUMBAI,
}

type NetworkToChainIds = {
  [key: string]: string
}

export const NETWORK_TO_CHAIN_IDS: NetworkToChainIds = {
  [ALL_NETWORKS.LOCALHOST]: '1337',
  [ALL_NETWORKS.SEPOLIA]: '11155111',
  [ALL_NETWORKS.MUMBAI]: '80001',
}

type ChainCryptos = {
  [key: string]: string
}

export const CHAIN_CRYPTOS: ChainCryptos = {
  '1337': 'ETH',
  '11155111': 'ETH',
  '80001': 'MATIC',
}
