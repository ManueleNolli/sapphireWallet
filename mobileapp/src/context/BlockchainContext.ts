import { createContext } from 'react'
import { NETWORKS } from '../constants/Networks'

type BlockchainContextType = {
  currentNetwork: NETWORKS
  setCurrentNetwork: (network: NETWORKS) => void
  ethersProvider: any
  setEthersProvider: (network: NETWORKS) => Promise<void>
}

export const BlockchainContext = createContext<BlockchainContextType>(<
  BlockchainContextType
>{
  currentNetwork: NETWORKS.LOCALHOST,
  setCurrentNetwork: (network: NETWORKS): void => {},
  ethersProvider: null,
  setEthersProvider: async (network: NETWORKS): Promise<void> => {},
})
