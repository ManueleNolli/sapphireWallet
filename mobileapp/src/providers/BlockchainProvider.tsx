import React, { useContext, useEffect } from 'react'
import { Provider } from 'ethers'
import useLoading from '../hooks/useLoading'
import { BlockchainContext } from '../context/BlockchainContext'
import { getProvider } from '../services/blockchain'
import { NETWORKS } from '../constants/Networks'
import Loading from '../pages/Loading/Loading'
import Error from '../pages/Error/Error'

type BlockchainProviderProps = {
  children: React.ReactNode
}

export function BlockchainProvider({ children }: BlockchainProviderProps) {
  const { currentNetwork } = useContext(BlockchainContext)
  const { isLoading, setIsLoading, isError, setIsError } = useLoading(true)

  const [provider, setProvider] = React.useState<Provider | null>(null)
  const [currentNetworkLocal, setCurrentNetwork] =
    React.useState<NETWORKS>(currentNetwork)

  const setEthersProvider = async (network: NETWORKS) => {
    let connectedProvider
    try {
      connectedProvider = await getProvider(network)
      setCurrentNetwork(network)
    } catch (e) {
      setIsError(true)
      console.error('Error connecting to blockchain')
    }

    if (connectedProvider) {
      setProvider(connectedProvider)
    }
  }

  useEffect(() => {
    setEthersProvider(currentNetwork).then(() => setIsLoading(false))
  }, [])

  if (isLoading) {
    return <Loading text={'Connecting to blockchain...'} />
  }

  if (isError) {
    return <Error text={'Error connecting to blockchain'} />
  }

  return (
    <BlockchainContext.Provider
      value={{
        currentNetwork: currentNetworkLocal,
        ethersProvider: provider,
        setEthersProvider,
      }}
    >
      {children}
    </BlockchainContext.Provider>
  )
}
