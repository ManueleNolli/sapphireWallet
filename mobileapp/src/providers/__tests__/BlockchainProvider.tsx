import React from 'react'
import { act, fireEvent, waitFor } from '@testing-library/react-native'
import renderWithTheme from '../../TestHelper'
import { BlockchainProvider } from '../BlockchainProvider'
import { getProvider } from '../../services/blockchain'
import { Text } from 'react-native'
import { BlockchainContext } from '../../context/BlockchainContext'
import { Provider } from 'ethers'
import { NETWORKS } from '../../constants/Networks'

jest.mock('../../pages/Loading/Loading')
jest.mock('../../pages/Error/Error')
jest.mock('../../services/blockchain')

function MockComponent() {
  const { currentNetwork } = React.useContext(BlockchainContext)
  return (
    <div>
      <Text>{currentNetwork}</Text>
    </div>
  )
}

describe('BlockchainProvider', () => {
  it('should set default start up network', async () => {
    ;(getProvider as jest.Mock).mockResolvedValue({} as Provider)
    let tree: any

    await waitFor(async () => {
      tree = renderWithTheme(
        <BlockchainProvider>
          <MockComponent />
        </BlockchainProvider>
      )
    })

    expect(tree.getByText(NETWORKS.SEPOLIA)).toBeTruthy()
  })

  it('should catch error at start up', async () => {
    ;(getProvider as jest.Mock).mockRejectedValue('error')
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {})
    let tree: any

    await waitFor(async () => {
      tree = renderWithTheme(
        <BlockchainProvider>
          <MockComponent />
        </BlockchainProvider>
      )
    })

    expect(spy).toHaveBeenCalledWith('Error connecting to blockchain')
    expect(tree).toMatchSnapshot()
  })

  it('render loading', async () => {
    ;(getProvider as jest.Mock).mockReturnValue(Promise.resolve())
    const tree = renderWithTheme(
      <BlockchainProvider>
        <MockComponent />
      </BlockchainProvider>
    )

    expect(tree).toMatchSnapshot()

    // Wait for loading
    await waitFor(() => {
      Promise.resolve()
    })
  })
})
