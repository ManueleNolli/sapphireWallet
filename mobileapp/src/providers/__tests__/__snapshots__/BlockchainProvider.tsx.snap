// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockchainProvider render loading 1`] = `
<div
  data-testid="Loading"
>
  Loading
</div>
`;

exports[`BlockchainProvider should catch error at start up 1`] = `
<div
  data-testid="Error"
>
  Error
</div>
`;
